components/chat_ui/browser/chat_ui_api_request.cc

```diff
@@ -0,0 +1,119 @@
+/* Copyright (c) 2022 The Brave Authors. All rights reserved.
+ * This Source Code Form is subject to the terms of the Mozilla Public
+ * License, v. 2.0. If a copy of the MPL was not distributed with this file,
+ * You can obtain one at https://mozilla.org/MPL/2.0/. */
+
+#include "brave/components/chat_ui/browser/chat_ui_api_request.h"
+
+#include <base/containers/flat_map.h>
+
+#include <utility>
+
+#include "base/json/json_writer.h"
+#include "brave/components/chat_ui/browser/buildflags.h"
+#include "brave/components/chat_ui/common/chat_ui_constants.h"
+#include "brave/components/constants/brave_services_key.h"
+#include "net/traffic_annotation/network_traffic_annotation.h"
+#include "services/network/public/cpp/shared_url_loader_factory.h"
+#include "url/gurl.h"
+
+namespace {
+
+net::NetworkTrafficAnnotationTag GetNetworkTrafficAnnotationTag() {
+  return net::DefineNetworkTrafficAnnotation("chat_ui", R"(
+      semantics {
+        sender: "ChatUI"
+        description:
+          "This is used to communicate with our partner API"
+          "on behalf of the user interacting with the ChatUI."
+        trigger:
+          "Triggered by user sending a prompt."
+        data:
+          "Will generate a text completion that attempts to match whatever context or pattern the user gave it"
+        destination: WEBSITE
+      }
+      policy {
+        cookies_allowed: NO
+        policy_exception_justification:
+          "Not implemented."
+      }
+    )");
+}
+
+GURL GetURLWithPath(const std::string& host, const std::string& path) {
+  return GURL(std::string(url::kHttpsScheme) + "://" + host).Resolve(path);
+}
+
+std::string CreateJSONRequestBody(base::ValueView node) {
```

Comment:
is this following the rule of 2? https://chromium.googlesource.com/chromium/src/+/master/docs/security/rule-of-2.md

Reply:
This was a paste from VPN's api request handler. Looks like VPN and AI Chat's handler parses JSON in the browser process which violates the rule.

Reply:
Ah, `CreateJSONRequestBody` isn't called for input that comes from the internet. It's only called for a stack-allocated base::Value::Dict. I think this should be ok?

Reply:
@bridiver can you verify if this is ok?